### PR TITLE
IngressRoute: add an option to disable cross-namespace routing

### LINKED ISSUE #7353: IngressRoute resource allows cross-namespace routing by default
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

This is a replication and investigation report for the issue reported in https://github.com/containous/traefik/issues/7151 and discussed on the community forums https://community.traefik.io/t/cross-namespaces-ingressroutes-and-services/7419/19

- Prior to Traefik v2.1, `IngressRoute` was not capable of routing to services outside of its own "root" namespace
- This limitation appears to coincide with the constraints on K8S native `Ingress` based on an [issue reported](https://github.com/containous/traefik/issues/5748#issuecomment-547323939) running Traefik 2.0.4
- This limitation was removed with the introduction of [this PR](https://github.com/traefik/traefik/pull/5711), and shipped with Traefik v2.1 to current.
- While a user can [define which namespaces can be watched](https://doc.traefik.io/traefik/providers/kubernetes-crd/#namespaces), this does not constrain the user's ability to restrict `IngressRoute` to their respective namespaces, which is based on an expectation set by native K8S `Ingress` objects and the behavior of `IngressRoute` prior to 2.1.
- This behavior is specific to Traefik and `IngressRoute` and I could reproduce the regressed behavior on v2.0, and produce the cross-namespace behavior on 2.3 on K8S 1.16, 1.17, and 1.18 (this is not something K8S will restrict)

### What did you expect to see?

`IngressRoute` provider configuration to contain a flag that enables cross-namespace routing behavior (this is a worthwhile feature, IMO).

### What did you see instead?

By default, `IngressRoute` will allow users to cross-namespace boundaries even though the user would expect this behavior to be disallowed given current constraints on `Ingress`

### Output of `traefik version`: (_What version of Traefik are you using?_)

* Traefik 2.0.4
* Traefik 2.3

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

See original issue for environment / configuration. Additional tests were run on:

* Kubernetes 1.16
* Kubernetes 1.17

### If applicable, please paste the log output in DEBUG level (`--log.level=DEBUG` switch)
n/a

### PR DESCRIPTION
**Pull request by rtribotte**
### What does this PR do?

This PR adds the `allowCrossNamespace` option on the Kubernetes provider configuration to allow/disallow cross-namespace routing.
The `allowCrossNamespace` is defaulted to `true`.
When set to `false` it's not possible for an IngressRoute to reference a service in another namespace than its namespace.

### Motivation

fixes #7353

### More

- [x] Added/updated tests
- [x] Added/updated documentation

### Additional Notes

Co-authored-by: Jean-Baptiste Doumenjou <925513+jbdoumenjou@users.noreply.github.com>

### CODE REVIEW

File: pkg/provider/kubernetes/crd/kubernetes.go [hunk 1]
```diff
@@ -43,6 +43,7 @@ type Provider struct {
 	CertAuthFilePath       string          `description:"Kubernetes certificate authority file path (not needed for in-cluster client)." json:"certAuthFilePath,omitempty" toml:"certAuthFilePath,omitempty" yaml:"certAuthFilePath,omitempty"`
 	DisablePassHostHeaders bool            `description:"Kubernetes disable PassHost Headers." json:"disablePassHostHeaders,omitempty" toml:"disablePassHostHeaders,omitempty" yaml:"disablePassHostHeaders,omitempty" export:"true"`
 	Namespaces             []string        `description:"Kubernetes namespaces." json:"namespaces,omitempty" toml:"namespaces,omitempty" yaml:"namespaces,omitempty" export:"true"`
+	EnableCrossNamespaces  *bool           `description:"Kubernetes enable cross namespaces capacity." json:"enableCrossNamespaces,omitempty" toml:"enableCrossNamespaces,omitempty" yaml:"enableCrossNamespaces,omitempty" export:"true"`
```
Comment (jspdown) [inline pkg/provider/kubernetes/crd/kubernetes.go:46]:
```suggestion
	EnableCrossNamespaces  *bool           `description:"Kubernetes enable cross namespaces capability." json:"enableCrossNamespaces,omitempty" toml:"enableCrossNamespaces,omitempty" yaml:"enableCrossNamespaces,omitempty" export:"true"`
```

**Review by jspdown — CHANGES_REQUESTED** [review]

**Comment by jspdown** [comment]
@jbdoumenjou, @rtribotte Could you add a new integration test?

File: docs/content/providers/kubernetes-crd.md [hunk 2]
```diff
@@ -250,6 +250,34 @@ providers:
...
+--providers.kubernetescrd.enableCrossNamespace=false
+```
+
+If the parameter is set to `false`, an IngressRoute will not be able to reference any service
+in another namespace than the IngressRoute namespace.
+
+!!! warning "Deprecation"
+    
+    Please notice that the default value for this option will be set to `false` in the next minor version (v2.4).
+
```
[the remark this answers is not in the document]
Comment (SantoDE) [reply docs/content/providers/kubernetes-crd.md:280]: Is 2.4 correct? Like, will we do another PR to master before releasing 2.4?

File: pkg/provider/kubernetes/crd/kubernetes.go [hunk 3]
```diff
@@ -98,6 +105,10 @@ func (p *Provider) Provide(configurationChan chan<- dynamic.Message, pool *safe.
 		return err
 	}
 
+	if k8sClient.enableCrossNamespaces {
+		logger.Warn("Cross-namespace referencing between Ingresses and Services is enabled, please ensure that is expected (see EnableCrossNamespaces option).")
```
Comment (SantoDE) [inline pkg/provider/kubernetes/crd/kubernetes.go:109]:
```suggestion
logger.Warn("Cross-namespace referencing between Ingresses and Services is enabled, please ensure that this is expected (see EnableCrossNamespaces option).")
```

**Review by SantoDE — CHANGES_REQUESTED** [review]
just some minor stuff :)

File: docs/content/providers/kubernetes-crd.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbdoumenjou) [reply docs/content/providers/kubernetes-crd.md:280]: @jspdown, it seems to be a good idea, we can enhance the documentation.
@SantoDE,  the plan was to introduce the flag without modifying the behavior, then reverse the flag to be more strict by default.

[the remark this answers is not in the document]
Comment (SantoDE) [reply docs/content/providers/kubernetes-crd.md:280]: Ok but in which versions? sorry, I dont get it 🙈

[the remark this answers is not in the document]
Comment (jbdoumenjou) [reply docs/content/providers/kubernetes-crd.md:280]: After discussions, we will allow cross namespace by default in this version and wait for user feedback.
We will reverse the flag in a future version.

File: pkg/provider/kubernetes/crd/kubernetes_test.go [hunk 4]
```diff
@@ -3592,3 +3610,562 @@ func TestGetServicePort(t *testing.T) {
...
+		{
+			desc:  "TCP cross namespace disallowed",
+			paths: []string{"tcp/services.yml", "tcp/with_cross_namespace.yml"},
+			expected: &dynamic.Configuration{
+				UDP: &dynamic.UDPConfiguration{
+					Routers:  map[string]*dynamic.UDPRouter{},
+					Services: map[string]*dynamic.UDPService{},
+				},
+				TCP: &dynamic.TCPConfiguration{
+					// The router that reference the invalid service will be discarded upper in the process
```
Comment (jspdown) [inline pkg/provider/kubernetes/crd/kubernetes_test.go:4024]:
```suggestion
					// The router that references the invalid service will be discarded upper in the process
```

File: pkg/provider/kubernetes/crd/kubernetes_test.go [hunk 5]
```diff
@@ -3592,3 +3610,562 @@ func TestGetServicePort(t *testing.T) {
...
+					Services: map[string]*dynamic.TCPService{},
+				},
+				TLS: &dynamic.TLSConfiguration{},
+			},
+		},
+		{
+			desc:  "UDP cross namespace disallowed",
+			paths: []string{"udp/services.yml", "udp/with_cross_namespace.yml"},
+			expected: &dynamic.Configuration{
+				// The router that reference the invalid service will be discarded upper in the process
```
Comment (jspdown) [inline pkg/provider/kubernetes/crd/kubernetes_test.go:4087]:
```suggestion
				// The router that references the invalid service will be discarded upper in the process
```

File: pkg/provider/kubernetes/crd/kubernetes_udp.go [hunk 6]
```diff
@@ -78,6 +78,10 @@ func (p *Provider) loadIngressRouteUDPConfiguration(ctx context.Context, client
 func createLoadBalancerServerUDP(client Client, namespace string, service v1alpha1.ServiceUDP) (*dynamic.UDPService, error) {
```
Comment (jspdown) [inline pkg/provider/kubernetes/crd/kubernetes_udp.go:78]: Would be nice to rename the `namespace` parameter into `parentNamespace` like you did in the createLoadBalancerServerTCP`. It's easier to understand where the namespace comes from.

**Review by jspdown — CHANGES_REQUESTED** [review]

**Review by kevinpollet — CHANGES_REQUESTED** [review]

**Review by jbdoumenjou — APPROVED** [review]
LGTM

**Review by jspdown — APPROVED** [review]
LGTM :+1:

**Review by kevinpollet — APPROVED** [review]
LGTM 👍

**Review by ldez — APPROVED** [review]
LGTM

**Pull request merged**